### PR TITLE
CAF-2613 Tenant index field type mappings

### PR DESCRIPTION
**Pull request by CJComac**

### CODE REVIEW

**Comment by glucy** [comment]
@CJComac There are currently conflicts.  Please resolve.  I will review in due course.

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java [hunk 1]
```diff
@@ -18,12 +18,18 @@
...
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.ResourceAlreadyExistsException;
 import org.elasticsearch.action.admin.indices.create.CreateIndexRequest;
 import org.elasticsearch.client.transport.TransportClient;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.*;
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
```
Comment (glucy) [inline caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Organise imports.
Remove * import, use explicits

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java [hunk 2]
```diff
@@ -80,6 +89,34 @@ private void createIndex(String indexName){
...
+        String cafAuditEventTenantIndexMappingsFileName = "CafAuditEventTenantIndexMappings.json";
+        //  Get the contents of the index mapping file and assign to byte array before attempting to parse JSON
+        byte[] cafAuditEventTenantIndexMappingsBytes;
+        try {
+            cafAuditEventTenantIndexMappingsBytes = ByteStreams.toByteArray(getClass().getClassLoader()
+                    .getResourceAsStream(cafAuditEventTenantIndexMappingsFileName));
+        } catch (IOException e) {
+            String errorMessage = "Unable to read bytes from " + cafAuditEventTenantIndexMappingsFileName;
+            LOG.error(errorMessage);
+            throw new RuntimeException(e);
```
Comment (glucy) [inline caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Add message and re-throw exception.

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java (hunk 1 above)
Comment (CJComac) [reply caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Cleaned up the imports in the latest commit

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java (hunk 2 above)
Comment (CJComac) [reply caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Adjusted in the latest commit

**Comment by CJComac** [comment]
@glucy I made your feedback changes and merged develop into the branch

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java [hunk 3]
```diff
@@ -33,6 +41,8 @@
     private final LoadingCache<String, String> indexCache;
     private final ElasticAuditConfiguration config;
 
+    private static final String indexTypeName = "cafAuditEvent";
```
Comment (glucy) [inline caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Add this to the constants file class

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java [hunk 4]
```diff
@@ -80,6 +91,34 @@ private void createIndex(String indexName){
                 .build();
         CreateIndexRequest indexRequest = new CreateIndexRequest(indexName, indexSettings);
 
+        String cafAuditEventTenantIndexMappingsFileName = "CafAuditEventTenantIndexMappings.json";
```
Comment (glucy) [inline caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Place "CafAuditEventTenantIndexMappings.json" in the constants file also.

File: caf-audit/src/test/java/com/hpe/caf/auditing/elastic/ElasticAuditIT.java [hunk 5]
```diff
@@ -19,10 +19,12 @@
...
 import com.hpe.caf.auditing.AuditEventBuilder;
+import org.elasticsearch.action.admin.cluster.state.ClusterStateResponse;
 import org.elasticsearch.action.admin.indices.delete.DeleteIndexRequest;
 import org.elasticsearch.action.delete.DeleteResponse;
 import org.elasticsearch.action.search.SearchType;
 import org.elasticsearch.client.transport.TransportClient;
+import org.elasticsearch.common.compress.CompressedXContent;
 import org.elasticsearch.index.query.QueryBuilders;
 import org.elasticsearch.rest.RestStatus;
 import org.elasticsearch.search.SearchHit;
```
Comment (glucy) [inline caf-audit/src/test/java/com/hpe/caf/auditing/elastic/ElasticAuditIT.java:30]: Replace import java.util.*; with individual import.
You should do this directly in develop and pull it into your branch.

We should always look to improve :-)

File: caf-audit/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java [hunk 6]
```diff
@@ -38,9 +39,12 @@
 import java.util.Date;
 import java.util.Map;
 import java.util.UUID;
+import java.util.concurrent.ExecutionException;
 
 public class GeneratedAuditLogIT {
 
+    private static final String testTenant = "tenant1";
+
     private static final Logger LOG = LoggerFactory.getLogger(GeneratedAuditLogIT.class);
```
Comment (glucy) [inline caf-audit/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java]: LOG is never used.  I suggest we use it.
If not used it should be removed, but I'd rather see it used.  @CJComac use it where appropriate in your tests, you don't need to retro-fit it.

File: caf-audit/src/test/java/com/hpe/caf/auditing/elastic/RetryElasticsearchOperation.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hpe.caf.auditing.elastic;
+
+public class RetryElasticsearchOperation
```
Comment (mulhollandc) [inline caf-audit/src/test/java/com/hpe/caf/auditing/elastic/RetryElasticsearchOperation.java]: Should this be renamed to ElasticAuditRetryOperation to be consistent with other classes under elastic?

File: caf-audit/src/test/java/com/hpe/caf/auditing/elastic/ElasticAuditIT.java (hunk 5 above)
Comment (CJComac) [reply caf-audit/src/test/java/com/hpe/caf/auditing/elastic/ElasticAuditIT.java:30]: Committed to develop and merged develop into this branch

File: caf-audit/src/test/java/com/hpe/caf/auditing/elastic/RetryElasticsearchOperation.java (hunk 7 above)
Comment (CJComac) [reply caf-audit/src/test/java/com/hpe/caf/auditing/elastic/RetryElasticsearchOperation.java]: Renamed in the latest commit.

File: caf-audit/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java (hunk 6 above)
Comment (CJComac) [reply caf-audit/src/test/java/com/hpe/caf/auditing/elastic/GeneratedAuditLogIT.java]: Removed in the latest commit. Logger is not used in ElasticAuditIT.

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java (hunk 4 above)
Comment (CJComac) [reply caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Moved into constants file in the latest commit.

File: caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java (hunk 3 above)
Comment (CJComac) [reply caf-audit/src/main/java/com/hpe/caf/auditing/elastic/ElasticAuditIndexManager.java]: Using the value in the constants file that holds the index type name in the latest commit.

**Review by glucy — APPROVED** [review]

**Pull request merged**